### ISSUE TITLE
[RFC] MoE weight format for transformers v5 models

### ISSUE THREAD

**Issue by piyifan123** [ByteDance-Seed/VeOmni#499]
# Background

In https://github.com/ByteDance-Seed/VeOmni/blob/main/docs/transformers_v5/transformers_v5_moe_weight_loading.md, there is a survey section about the HF safetensor expert weight format and the HF modeling code.

## transformer-v5 expert modeling format

In short, for models added after transformers v5 (the most notable example is qwen3.5), the expert weight in the safetensor and modeling code is set to the following format:

```
model.language_model.layers.0.mlp.experts.gate_up_proj  [num_experts, 2 * I, H]
model.language_model.layers.0.mlp.experts.down_proj     [num_experts, H, I]

self.gate_up_proj = nn.Parameter(torch.empty(self.num_experts, 2 * self.intermediate_dim, self.hidden_dim))
self.down_proj = nn.Parameter(torch.empty(self.num_experts, self.hidden_dim, self.intermediate_dim))
```

## HF weight loading behavior

HuggingFace has also modified its modeling code to align all MoE models (including old models added before v5) to align with the above format. To keep the new code backward compatible with safetensor uploaded before, HuggingFace decided to implement a runtime conversion:

- core conversion code: https://github.com/huggingface/transformers/blob/main/src/transformers/core_model_loading.py#L989
- conversion rules https://github.com/huggingface/transformers/blob/main/src/transformers/conversion_mapping.py

## veomni weight loading behavior before

To support fused MoE, veomni basically just changes the modeling to take the following format:
```
        self.gate_proj = nn.Parameter(torch.empty(self.num_experts, self.intermediate_dim, self.hidden_dim))
        self.up_proj = nn.Parameter(torch.empty(self.num_experts, self.intermediate_dim, self.hidden_dim))
        self.down_proj = nn.Parameter(torch.empty(self.num_experts, self.hidden_dim, self.intermediate_dim))
```

(The difference from transformers v5 is mostly gate and up are separate tensors instead of a stacked one).

Instead of having a weight loading logic to handle this (like HF), veomni asks users to convert the safetensor checkpoint before training via https://github.com/ByteDance-Seed/VeOmni/tree/main/scripts/moe_ckpt_merge to match the above format.

(For recently [merged](https://github.com/ByteDance-Seed/VeOmni/pull/496) qwen3-moe transformers v5 support, we keep the above behavior (i.e., patch the experts module to use the veomni moe weight format))

# Decision to make 1: modeling weight format for transformers v5 models

We need to agree on the weight format in veomni modeling code for transformers v5:

- Option 1: match the transformers v5 format: `up_gate: [E, 2 * I, H], down: [E, H, I]`:
   -  (+) No more special handling for new MoE models like Qwen3.5 going forward.
   -  (-) Requires a new fused moe implementation to match the format (not that hard, just need to fuse up/gate into one group gemm).
   -  (-) Does not match the ByteDance internal fused moe interface. 
- Option 2: keep the current veomni format
   -  (+) keeps the same moe interface.
   -  (-) ongoing headaches for future new OSS MoE models which will likely follow the transformers v5 format. 
       -  We note that one exception: glm_moe_dsa (GLM5 architecture) which is [added](https://github.com/huggingface/transformers/pull/43858) after v5 but still uses per expert split format in safetensor. The HF maintainer seems not to enforce newly introduced models to match their new format in safetensors...  

# Decision to make 2: load mismatch safetensor

No matter which option taken above, we will have safetensor mismatch from the modeling code anyway (e.g., if we take Option 1, all old OSS models weights have a mismatch. If we take Option 2, all old and new model weights will have a mismatch). So we need a decision on how to load safetensor for mismatch models:

- Option A: Still requires users to convert to the new format offline.
  - (+) offline conversion is generally safer and easier to test than runtime conversion.
  - (-) Confusions and footgun for users who forget to convert. Though there are workaorunds:
      - we can add new runtime validation to warn users. 
      - we can also upload converted safetensors to HF.
  - (-) The training outcome ckpt needs another split to match the safetensor format.  
- Option B: Implement runtime loading conversion.
  - (+) no more prerequisite for users to convert the model ckpt. 
  - (-) more subtle and hard-to-debug bugs for the dynamic loading logic. For example, https://github.com/huggingface/transformers/blob/main/src/transformers/core_model_loading.py#L989 the HF logic is very complex which shows that our logic can be this complex.
  - (-) DCP ckpt still uses the new format (since it just saves the nn.Module hierarchy) which means an offline conversion back to the legacy format is still needed for training outcome ckpt even we adopt this approach.

# Caveats: verl-veomni integration

There are 2 places verl-veomni integration need to handle this:
1. Initial model loading from HF safetensor
    -  It just calls the veomni `build_foundation_model` API which shares the same weight loading behavior as veomni. Ref: https://github.com/verl-project/verl/blob/main/verl/workers/engine/veomni/utils.py#L103
2. When transfer weight from training engine to LLM inference server:
     - There is a per model mapping logic to map back to Safetensor format (vllm/sglang assumes the safetensor format as weight loading input) https://github.com/verl-project/verl/blob/main/verl/workers/engine/veomni/utils.py#L103

**Comment by piyifan123** [comment]
After discussions with @FoolPlayer @Coach257 @Luosuu @TimYangst, our decision is

1. Option 1 chosen: match the transformers v5 format: up_gate: [E, 2 * I, H], down: [E, H, I]
   - We need to add an option in our fused moe op to take [E, 2*I, H] concat gate-up weights to avoid forward time reshaping and .contiguous() call. https://github.com/ByteDance-Seed/VeOmni/blob/main/veomni/ops/fused_moe/group_gemm.py#L269-L277
2. Option B chosen: Implement runtime loading conversion.
   - we have heard a lot of complaints about requiring pre-merged weights to train and it also confuses veRL community when they use veomni as the training engine.
   - Implementation wise, instead of implementing [a complex rule based system](https://github.com/huggingface/transformers/blob/main/src/transformers/conversion_mapping.py#L69) and [a rule execution engine](https://github.com/huggingface/transformers/blob/11b1906d5c0dae39c13270e47cc02c4cde70e548/src/transformers/core_model_loading.py#L168) like transformers, we will implement a per model conversion tool that use pure pytorch code to transform so that it's easy to understand and test.
